### PR TITLE
resolve f-string warnings

### PR DESCRIPTION
**Pull request by jenningsje**

### CODE REVIEW

File: src/sos/__main__.py [hunk 1]
```diff
@@ -564,8 +563,7 @@ def cmd_run(args, workflow_args):
     try:
         # workflow args has to be in the format of --arg, not positional, not -a
         if workflow_args and not workflow_args[0].startswith("--"):
-            raise ValueError("Unrecognized command line option {}".format(
-                " ".join(workflow_args)))
+            raise ValueError(f"Unrecognized command line option {" ".join(workflow_args)}")
```
Comment (BoPeng) [inline src/sos/__main__.py:566]: This will not work due to the nested use of double quote. Use a different quote inside.

**Comment by BoPeng** [comment]
There is no need to create and push a commit for every single change. It is a huge waste of time.

File: src/sos/__main__.py [hunk 2]
```diff
@@ -1156,11 +1153,9 @@ def preview_file(previewers, filename, style=None):
             "metadata": {},
             "data": {
                 "text/plain":
-                    "\n> {} ({}):".format(
-                        filename, pretty_size(os.path.getsize(filename))),
+                    f"\n> {filename} ({pretty_size(os.path.getsize(filename))}):",
                 "text/html":
-                    HTML('<div class="sos_hint">> {} ({}):</div>'.format(
-                        filename, pretty_size(os.path.getsize(filename)))).data,
+                    HTML(f'<div class="sos_hint">> {filename} ({pretty_size(os.path.getsize(filename))}').data,
```
Comment (BoPeng) [inline src/sos/__main__.py:1158]: Missing `</div>`.

File: src/sos/__main__.py [hunk 3]
```diff
@@ -1332,12 +1326,11 @@ def cmd_preview(args, unknown_args):
                     print(msg[1]["data"]["text/plain"])
                 elif "text/html" in msg[1]["data"]:
                     print(msg[1]["data"]["text/html"])
-                else:
-                    print("BINARY DATA of type {}".format(", ".join(
-                        msg[1]["data"].keys())))
+                if "text/plain" in msg[1]["data"]:
+                    print(msg[1]["data"]["text/plain"])
```
Comment (BoPeng) [inline src/sos/__main__.py:1330]: This change does not match the original code.

File: src/sos/__main__.py [hunk 4]
```diff
@@ -1546,15 +1538,15 @@ def cmd_execute(args, workflow_args):
...
+                    env.logger.warning(f"{t} ``{s}``")
                     failed_tasks.add(t)
             if all(x in ("completed", "failed", "aborted") for x in res):
-                raise RuntimeError(
-                    "{} completed, {} failed, {} aborted)".format(
-                        len([x for x in res if x == "completed"]),
-                        len([x for x in res if x == "failed"]),
-                        len([x for x in res if x.startswith("aborted")]),
-                    ))
+                into_runtime_error = (
```
Comment (BoPeng) [inline src/sos/__main__.py:1544]: what is this?

File: src/sos/actions.py [hunk 5]
```diff
@@ -773,13 +768,14 @@ def sos_run(workflow=None,
...
                 "SOS_DEBUG"]:
-            env.log_to_file(
-                "ACTION",
-                "Executing workflow ``{}`` with input ``{}`` and {}".format(
-                    workflow,
-                    short_repr(env.sos_dict.get("_input", None), True),
-                    "no args" if not args_output else args_output,
-                ),
+
+            into_log_to_file = (
```
Comment (BoPeng) [inline src/sos/actions.py:772]: why this?

File: src/sos/parser.py [hunk 6]
```diff
@@ -1474,11 +1473,10 @@ def workflow(self,
...
                 raise ValueError(
-                    "Name of workflow should be specified because "
-                    "the script defines more than one pipelines without a default one. "
-                    "Available pipelines are: {}.".format(", ".join(
-                        self.workflows)))
-        elif wf_name not in self.workflows and wf_name != "default":
+                    f"Name of workflow should be specified because "
+                    f"the script defines more than one pipelines without a default one. "
+                    f"Available pipelines are: {', '.join(self.workflows)}.")
+        if wf_name not in self.workflows and wf_name != "default":
```
Comment (BoPeng) [inline src/sos/parser.py:1479]: You again changed the logic of the code.

File: src/sos/tasks.py [hunk 7]
```diff
@@ -1685,26 +1685,26 @@ def row(th=None, td=None):
     elif verbosity == 0:
         print("\n".join(obtained_status))
     elif verbosity == 1:
-        for s, (t, d) in zip(obtained_status, all_tasks):
+        for s, (t, dr) in zip(obtained_status, all_tasks):
```
Comment (BoPeng) [inline src/sos/tasks.py:1688]: what is the reason for this change?

File: src/sos/tasks.py [hunk 8]
```diff
@@ -1685,26 +1685,26 @@ def row(th=None, td=None):
...
             print(f"{t}\t{s}")
     elif verbosity == 2:
         tsize = 20
-        for s, (t, d) in zip(obtained_status, all_tasks):
+        for s, (t, dr) in zip(obtained_status, all_tasks):
```
Comment (BoPeng) [inline src/sos/tasks.py:1692]: This one?

File: src/sos/actions.py (hunk 5 above)
Comment (BoPeng) [reply src/sos/actions.py:772]: I see what you are trying to do but `env.log_to_file` has two parematers....

File: src/sos/workflow_engines.py [hunk 9]
```diff
@@ -182,17 +188,8 @@ def purge_workflows(self,
...
-                    self.agent.config.get("sos", "sos"),
-                    " ".join(workflows),
-                    "--all" if purge_all else "",
-                    f"--age {age}" if age is not None else "",
-                    f'--status {" ".join(status)}'
-                    if status is not None else "",
-                    f'--tags {" ".join(tags)}' if tags is not None else "",
-                    verbosity,
-                ))
+            return self.agent.check_output(check_output2
```
Comment (BoPeng) [inline src/sos/workflow_engines.py:191]: You cannot move the definition of the string before the parameters are used.

File: src/sos/__main__.py (hunk 4 above)
Comment (jenningsje) [reply src/sos/__main__.py:1544]: It is a variable that stores all of the f strings so that they do not have to be contained in a single line of code within the RuntimeError.

File: src/sos/tasks.py (hunk 8 above)
Comment (jenningsje) [reply src/sos/tasks.py:1692]: I concluded that d was a typo here as well since it was an unused variable. Should I revert it back to the original format?

File: src/sos/tasks.py (hunk 7 above)
Comment (jenningsje) [reply src/sos/tasks.py:1688]: There was a warning that stated that d (line 1688) was an unused variable. I concluded that it was a typo. Should I revert it back to d?

File: src/sos/__main__.py (hunk 3 above)
Comment (jenningsje) [reply src/sos/__main__.py]: I reverted the changes recently.

File: src/sos/__main__.py (hunk 2 above)
Comment (jenningsje) [reply src/sos/__main__.py]: I reverted f'<div back to '<div. These changes can be found in the most recent commit.

File: src/sos/actions.py (hunk 5 above)
Comment (jenningsje) [reply src/sos/actions.py:772]: I resolved this very recently. env.log_to_file now takes two arguments: "ACTION" and into_log_to_file.

File: src/sos/workflow_engines.py [hunk 10]
```diff
@@ -182,17 +178,20 @@ def purge_workflows(self,
...
+            check_output2 = (
+                f"{self.agent.config.get('sos', 'sos')} purge"
+                f"{' '.join(workflows)}"
+                f"{'--all' if purge_all else ''}"
+                f"--age {age}" if age is not None else ''
+                f"--status {' '.join(status)}"
+                if status is not None else ''
+                f"--tags {' '.join(tags) if tags is not None else ''}"
+                f"{verbosity}"
+            )
```
Comment (BoPeng) [inline src/sos/workflow_engines.py:191]: Missing `-v`

Comment (jenningsje) [reply src/sos/workflow_engines.py:191]: I'll begin resolving this issue right now.

File: src/sos/workflow_engines.py (hunk 9 above)
Comment (jenningsje) [reply src/sos/workflow_engines.py:192]: I recently submitted a commit where check_output2 is under 'try:' in purge_workflows and between return.self.agent.check_output().

File: src/sos/tasks.py (hunk 8 above)
Comment (jenningsje) [reply src/sos/tasks.py]: I just submitted a commit that replaces the variable dr with an underscore for lines 1688, 1692 and 1699.

File: src/sos/tasks.py (hunk 7 above)
Comment (jenningsje) [reply src/sos/tasks.py]: I just checked the latest commit on my GitHub page. If you navigate to https://github.com/jenningsje/Vat_sos/blob/consider-using-f-string/src/sos/tasks.py, you will find that dr has been replaced with underscore for lines 1688, 1692, 1699, 1707.

File: src/sos/workflow_engines.py (hunk 10 above)
Comment (jenningsje) [reply src/sos/workflow_engines.py:190]: I just checked my GitHub page. -v has now been added to line 188 within src/sos/workflow_engines.py.

File: src/sos/parser.py (hunk 6 above)
Comment (jenningsje) [reply src/sos/parser.py:1479]: I just checked my latest commit on https://github.com/jenningsje/Vat_sos/blob/consider-using-f-string/src/sos/parser.py. The two f'strings in lines 1476 and 1477 above have been replaced with normal strings.

File: src/sos/__main__.py [hunk 11]
```diff
@@ -1156,11 +1153,9 @@ def preview_file(previewers, filename, style=None):
             "metadata": {},
             "data": {
                 "text/plain":
-                    "\n> {} ({}):".format(
-                        filename, pretty_size(os.path.getsize(filename))),
+                    f"\n> {filename} ({pretty_size(os.path.getsize(filename))}):",
                 "text/html":
-                    HTML('<div class="sos_hint">> {} ({}):</div>'.format(
-                        filename, pretty_size(os.path.getsize(filename)))).data,
+                    HTML('<div class="sos_hint">> {filename} ({pretty_size(os.path.getsize(filename))}').data,
```
Comment (BoPeng) [inline src/sos/__main__.py:1158]: Where is that `</div>`?

File: src/sos/__main__.py [hunk 12]
```diff
@@ -1275,7 +1269,7 @@ def cmd_preview(args, unknown_args):
 
         if "GENERAL" in env.config["SOS_DEBUG"] or "ALL" in env.config[
                 "SOS_DEBUG"]:
-            env.log_to_file("GENERAL", 'Running "{}"'.format(" ".join(rargs)))
+            env.log_to_file("GENERAL", f"Running {' '.join(rargs)}")
```
Comment (BoPeng) [inline src/sos/__main__.py:1272]: Where is the `"` around `{}`. You can use triple quote to avoid conflicts.

File: src/sos/__main__.py (hunk 4 above)
Comment (BoPeng) [reply src/sos/__main__.py:1544]: where are the "completed", "failed", "aborted" etc?

File: src/sos/workflow_engines.py [hunk 13]
```diff
@@ -182,17 +178,19 @@ def purge_workflows(self,
...
-                    f'--tags {" ".join(tags)}' if tags is not None else "",
-                    verbosity,
-                ))
+            check_output2 = (
+                f"{self.agent.config.get('sos', 'sos')} purge"
+                f"{' '.join(workflows)}"
+                f"{'--all' if purge_all else ''}"
+                f"--age {age}" if age is not None else ''
+                f"--status {' '.join(status)}"
+                if status is not None else ''
```
Comment (BoPeng) [inline src/sos/workflow_engines.py:187]: This will not work since the stuff before "if" and after "else" will become two big string.

File: src/sos/singularity/client.py [hunk 14]
```diff
@@ -350,10 +345,10 @@ def run(self,
...
             if ret != 0:
                 debug_script_dir = env.exec_dir
-                msg = 'The script has been saved to {}/{}. To reproduce the error please run:\n``{}``'.format(
-                    debug_script_dir, tempscript,
-                    cmd.replace(f'{path(tempdir):p}',
-                                f'{path(debug_script_dir):p}'))
+                msg = (
+                    f"The script has been saved to {debug_script_dir}/{debug_script_dir}."
+                    f"To reproduce the error please run:\n``{cmd.replace(f'{path(tempdir):p}', f'{path(debug_script_dir):p}')}``"
+                )
```
Comment (BoPeng) [inline src/sos/singularity/client.py:351]: Please define a variable for 
```cmd.replace(f'{path(tempdir):p}',
                                f'{path(debug_script_dir):p}')```

to avoid using nested f-string.

File: src/sos/task_engines.py [hunk 15]
```diff
@@ -612,17 +612,17 @@ def query_tasks(self,
...
+                f"status {'' if tasks is None else ' '.join(tasks)} -v {verbosity}"
+                f"{'--all tasks' if check_all else ''}"
+                f"{'--html' if html else ''}"
+                f"{'--numeric-times' if numeric_times else ''}"
+                f'--age {age}' if age else ''
+                f'--tags {" ".join(tags)}' if tags else ''
+                f'--status {" ".join(status)}' if status else ''
+            )
+
+
```
Comment (BoPeng) [inline src/sos/task_engines.py:625]: this does not work

File: src/sos/task_engines.py [hunk 16]
```diff
@@ -662,11 +662,12 @@ def kill_tasks(self, tasks, tags=None, all_tasks=False):
...
-            self.agent.config.get('sos',
-                                  'sos'), '' if all_tasks else ' '.join(tasks),
-            f'--tags {" ".join(tags)}' if tags else '',
-            '--all tasks' if all_tasks else '')
+        cmd = (
+            f"{self.agent.config.get('sos', 'sos')}"
+            f"kill {'' if all_tasks else ' '.join(tasks)}"
+            f"--tags {' '.join(tags)} if tags else"
+            f"--all tasks if all_tasks else ''"
+            )
```
Comment (BoPeng) [inline src/sos/task_engines.py:671]: same thing. The solution would be to create variables for the `if else` values.

File: src/sos/task_engines.py [hunk 17]
```diff
@@ -705,13 +706,13 @@ def purge_tasks(self,
...
-                    f'--status {" ".join(status)}' if status is not None else
-                    '', f'--tags {" ".join(tags)}' if tags is not None else '',
-                    verbosity))
+                f"{self.agent.config.get('sos', 'sos'), ' '.join(tasks)}"
+                f"{'--all' if purge_all else ''}"
+                f"--age {age if age is not None else ''}"
+                f"--status {' '.join(status) if status is not None else ''}"
+                f"--tags {' '.join(tags) if tags is not None else ''}"
+                f"{verbosity}"
+                )
```
Comment (BoPeng) [inline src/sos/task_engines.py:715]: same problem here.

File: src/sos/tasks.py [hunk 18]
```diff
@@ -1421,7 +1421,7 @@ def row(th=None, td=None):
                     f'<tr><th align="right"  width="30%">{th}</th><td align="left"><div class="one_liner">{td}</div></td></tr>'
                 )
 
-        for s, (t, d) in zip(obtained_status, all_tasks):
+        for s, (t, dr) in zip(obtained_status, all_tasks):
```
Comment (BoPeng) [inline src/sos/tasks.py:1424]: this should also be changed.

File: src/sos/utils.py [hunk 19]
```diff
@@ -1119,11 +1119,13 @@ def run(self):
...
             second_elapsed = time.time() - self.start_time
-            prog.set_description("Elapsed time {}{}".format(
-                "" if second_elapsed < 86400 else
-                f'{int(second_elapsed/86400)} day{"s" if second_elapsed > 172800 else ""} ',
-                time.strftime("%H:%M:%S", time.gmtime(second_elapsed)),
-            ))
+
+            into_set_description = (
+                "Elapsed time" '' if second_elapsed < 86400 else
+                f"{int(second_elapsed/86400)} day{'s' if second_elapsed > 172800 else ''} "
```
Comment (BoPeng) [inline src/sos/utils.py:1125]: same problem.

**Comment by jenningsje** [comment]
Commit 9c8de41 is changes the following file: src/sos/workflow_engines.py.

**Comment by jenningsje** [comment]
Comment b42172c resolves lines 496-497 and 483 for src/sos/docker/client.py.

**Comment by jenningsje** [comment]
Comment 500e1a7 resolves lines 622 and 667 in src/sos/task_engines.py.

**Comment by BoPeng** [comment]
replaced by #1483
